[WIP] Add circuit equivalence checker

chriseclectic opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Adds an identity checker mode to the `UnitarySimulator` that can be used as a basis for a circuit equivalence checker.

This mode is activated by using `backend_options={"identity_checker": True}` for the `UnitarySimulator` and in this case instead of returning the final unitary matrix the result will say if the final unitary is `exp(i * theta) * U_target` where `U_target` is the identity matrix, or a custom target unitary matrix. If result is `True` it also returns the global phase `theta`. 

### Details and comments

To manually activate the identity checking mode one can do the following

```python
# Get ideal circuit
target_circuit = ... # Target quantum circuit
target_unitary = qiskit.execute(circuit_target, UnitarySimulator()).result().get_unitary(target_circuit)

# Check new circuit
test_circuit = ... # new quantum circuit
opts = {
    "identity_checker": True, 
    "target_unitary": target_unitary
}
job = qiskit.execute(test_circuit, UnitarySimulator(),
    backend_options=opts)
results = job.results()
print(results.data(0))
# prints 
{"identity": [True, 0.0]}  # output is [pass, global_phase]
```

A more efficient method is to compares $U_{target}^\dagger . U = id$ (`inverse(target_circuit).circuit` to the identity):

```python
target_circuit = ...
test_circuit = ... 
check_circuit = target_circuit.inverse() + test_circuit
opts =  {"identity_checker": True }
job = qiskit.execute(check_circuit, UnitaryVerifier(),
    backend_options=opts)
results = job.results()
print(results.data(check_circuit))
```

### TODO

Decide how to expose this functionality to the user for a circuit equivalence checker such as `check_equivalent(circuit1, circuit2)`

atilag commented:
I like the idea of these simulators being backends of the `UnitarySimulator`, and use them by passing a backend option.

chriseclectic commented:
Revisiting this PR after the 0.2 release and updated the description. I have removed the separate backends so that now there is an identity checker mode for the unitary controller / simulator. How to expose this functionality to the end user still needs to be discussed.
